base/loading.jl

```diff
@@ -564,6 +898,10 @@ function create_expr_cache(input::String, output::String, concrete_deps::typeof(
                   empty!(Base._concrete_dependencies)
                   append!(Base._concrete_dependencies, $concrete_deps)
                   Base._track_dependencies[] = true
+                  const $ENVINFO = $envinfo
+                  # same horrible hack as in _require
+                  ccall(:jl_set_global, Cvoid, (Any, Any, Any),
+                        Base.__toplevel__, $(Meta.quot(ENVINFO)), $envinfo)
```

Comment:
This allows correct new-style module loading in the child precompile process. The failure is later, however, when the precompiled module needs to get loaded into the original parent process, which tries to load it into Main.